java/maven/src/org/netbeans/modules/maven/queries/MavenPrimingReloadImplementation.java

```diff
@@ -0,0 +1,251 @@
+     * unless "force" is in effect - they are unlikely to be retrieved, since they already failed.
+     */
+    // @GuardedBy(this)
+    private Set<String> fakeArtifactNames = new HashSet<>();
+
+    
+    private Reference<ProjectStateData> lastData = new WeakReference<>(null);
+    
+    static class LC {
+        boolean firstRun = true;
+    }
+    
+    // just pretend MavenProject is not there, do not make conflicts
+    static class H {
+        final MavenProject p;
+
+        public H(MavenProject p) {
+            this.p = p;
+        }
+    }
+
+    static long getTimestamp(MavenProject mp) {
+        return mp == null ? -1 : MavenProjectCache.getLoadTimestamp(mp);
+    }
+    
+    private ProjectReload.Quality getProjectQuality(MavenProject mp) {
+        if (MavenProjectCache.isFallbackproject(mp)) {
+            // could not load at all.
+            MavenProject fallback = MavenProjectCache.getPartialProject(mp);
+            return fallback != null ? 
+                    ProjectReload.Quality.BROKEN:
+                    ProjectReload.Quality.NONE;
+        }
+        if (!MavenProjectCache.getFakedArtifacts(mp).isEmpty()) {
+            return ProjectReload.Quality.LOADED;
+        }
+        return ProjectReload.Quality.RESOLVED;
+    }
+    
+    
+    private ProjectStateData createStateData(MavenProject mp) {
+        synchronized (this) {
+            ProjectStateData<H> d2 = lastData.get();
+            if (d2 != null) {
+                H h = d2.getProjectData();
+                if (h != null && h.p == mp) {
+                    return d2;
+                }
+            }
+        }
+        ProjectStateData d;
+        
+        ProjectStateBuilder builder = ProjectStateData.builder(getProjectQuality(mp)).
+                timestamp(MavenProjectCache.getLoadTimestamp(mp));
+        builder.data(new H(mp));
+        builder.attachLookup(Lookups.fixed(mp));
+        d = builder.build();
+        synchronized (this) {
+            ProjectStateData d2 = lastData.get();
+            if (d2 != null) {
+                if (d2.getProjectData() == mp) {
+                    return d2;
+                }
+                d2.fireChanged(true, false);
+            }
+            lastData = new WeakReference<>(d);
+        }
+        return d;
+    }
+
+    static boolean isProjectFullyLoaded(MavenProject p) {       
+        return p == null || !MavenProjectCache.isIncompleteProject(p) ||
+               MavenProjectCache.getFakedArtifacts(p).isEmpty();
+    }
+
+    static String artifactGav(Artifact a) {
+        return String.format("%s:%s:%s:%s", a.getGroupId(), a.getArtifactId(), a.getVersion(), a.getClassifier());
+    }
+
+    @NbBundle.Messages({
+        "# {0} - project name",
+        "ERR_PrimingBuildFailed=Priming build of {0} failed."
+    })
+    @Override
+    public CompletableFuture reload(Project project, ProjectReload.StateRequest request, LoadContext context) {
```

Comment:
* The API client may chain processing after the load completes / gets errors
* depending on the request properties, the reload need not to happen, if cached data is current (implementation may decide)
* can be invoked through API unlike 'reload pom' action
* reload POM will not download added artifacts that are not locally available; this API can do it (yuck, priming, yet a while)

"reload POM" action could be reimplemented using this API

Reply:
right. We should not forget to follow up on this. Rename the action to "reload project" perhaps, get rid of the reload-pom implementation. Since this sounds like there could be two implementations now with overlapping functionality.

Reply:
Will try to incorporate, if not easy or implies more changes, will create an additional PR.

Reply:
Reimplemented in 65680f2c8f34